File: lv1-library-loan/csharp/LibraryLoan.Domain/Books/Book.cs
using LibraryLoan.BuildingBlocks;

namespace LibraryLoan.Domain.Books;

public sealed class Book : Entity<BookId>
{
    public string Title { get; }
    public string Author { get; }
    public Isbn Isbn { get; }

    public Book(BookId id, string title, string author, Isbn isbn)
        : base(id)
    {
        if (string.IsNullOrWhiteSpace(title))
        {
            throw new ArgumentException("タイトルは空にできません。", nameof(title));
        }

        if (string.IsNullOrWhiteSpace(author))
        {
            throw new ArgumentException("著者は空にできません。", nameof(author));
        }

        Title = title;
        Author = author;
        Isbn = isbn;
    }
}

File: lv1-library-loan/typescript/src/domain/books/book.ts
import { Entity } from "../../building-blocks/entity.js";
import type { BookId } from "./book-id.js";
import type { Isbn } from "./isbn.js";

export class Book extends Entity<BookId> {
  readonly title: string;
  readonly author: string;
  readonly isbn: Isbn;

  constructor(id: BookId, title: string, author: string, isbn: Isbn) {
    super(id);

    if (title.trim().length === 0) {
      throw new Error("タイトルは空にできません。");
    }
    if (author.trim().length === 0) {
      throw new Error("著者は空にできません。");
    }

    this.title = title;
    this.author = author;
    this.isbn = isbn;
  }
}
